File: src/tutorial/go2_tutorial_py/go2_tutorial_py/go2_ctrl.py
"""
需求：以动态传参的方式控制 go2 运动
"""

import rclpy
from rclpy.node import Node
from unitree_api.msg import Request

import json
from .sport_model import ROBOT_SPORT_API_IDS


class Go2CtrlPy(Node):

    def __init__(self):
        super().__init__('go2_ctrl_py')
        self.get_logger().info('go2_ctrl_py 创建')

        # 声明参数
        self.declare_parameter('sport_api_id', ROBOT_SPORT_API_IDS["BALANCESTAND"])
        self.declare_parameter('x', 0.0)
        self.declare_parameter('y', 0.0)
        self.declare_parameter('z', 0.0)

        self.req_pub = self.create_publisher(Request, '/api/sport/request', 10)
        self.timer = self.create_timer(0.1, self.on_timer)

    def on_timer(self):
        request = Request()

        # 读取参数
        api_id = self.get_parameter('sport_api_id').get_parameter_value().integer_value
        request.header.identity.api_id = api_id
        if api_id == ROBOT_SPORT_API_IDS["MOVE"]:
            js = {
                "x": self.get_parameter('x').get_parameter_value().double_value,
                "y": self.get_parameter('y').get_parameter_value().double_value,
                "z": self.get_parameter('z').get_parameter_value().double_value
            }
            request.parameter = json.dumps(js)

        self.req_pub.publish(request)


def main(args=None):
    rclpy.init(args=args)
    node = Go2CtrlPy()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/tutorial/go2_tutorial/src/go2_ctrl.cpp
/*
需求：以动态传参的方式控制go2运动

*/

#include <rclcpp/rclcpp.hpp>
#include <unitree_api/msg/request.hpp>
#include <sport_model.hpp>
#include <nlohmann/json.hpp>

using namespace std::chrono_literals;


class Go2Ctrl: public rclcpp::Node{
public:
    Go2Ctrl(const std::string &node_name):Node(node_name){
      RCLCPP_INFO(this->get_logger(),"go2_ctrl 创建");

      // 设置参数
      this->declare_parameter("sport_api_id", ROBOT_SPORT_API_ID_BALANCESTAND);
      this->declare_parameter("x", 0.0);
      this->declare_parameter("y", 0.0);
      this->declare_parameter("z", 0.0);

      req_pub_ = this->create_publisher<unitree_api::msg::Request>("/api/sport/request",10);
      timer_ = this->create_wall_timer(100ms, std::bind(&Go2Ctrl::on_timer, this));

    }
private:
    void on_timer(){
      // 创建速度指令并发布
      unitree_api::msg::Request request;

      // 设置参数
      auto id = this->get_parameter("sport_api_id").as_int();
      request.header.identity.api_id = id; // 运动模式
      if(id == ROBOT_SPORT_API_ID_MOVE){
        nlohmann::json js;
        js["x"] = this->get_parameter("x").as_double();
        js["y"] = this->get_parameter("y").as_double();
        js["z"] = this->get_parameter("z").as_double();
        request.parameter = js.dump();
      }

      req_pub_->publish(request);
    }

    rclcpp::Publisher<unitree_api::msg::Request>::SharedPtr req_pub_;
    rclcpp::TimerBase::SharedPtr timer_;
};
 
int main(int argc, char* argv[]){
    rclcpp::init(argc, argv);
    auto node = std::make_shared<Go2Ctrl>("go2_ctrl");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}